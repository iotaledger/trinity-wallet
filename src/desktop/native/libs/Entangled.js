const fork = require('child_process').fork;
const path = require('path');
const EntangledNode = require('entangled-node');

let timeout = null;

/**
 * Spawn a child process and return result in async
 * @param {object} payload - Payload to send to the child process
 * @returns {Promise}
 */
const exec = (payload) => {
    return new Promise((resolve, reject) => {
        const child = fork(path.resolve(__dirname, 'Entangled.js'));

        const { job } = JSON.parse(payload);

        child.on('message', (message) => {
            resolve(message);

            clearTimeout(timeout);
            child.kill();
        });

        timeout = setTimeout(
            () => {
                reject(`Timeout: Entangled job: ${job}`);
                child.kill();
            },
            job === 'batchedPow' ? 180 * 1000 : 30 * 1000,
        );

        child.send(payload);
    });
};

/**
 * If module called as a child process, execute requested function and return response
 */
process.on('message', async (data) => {
    const payload = JSON.parse(data);

    if (payload.job === 'pow') {
        const pow = await EntangledNode.powTrytesFunc(payload.trytes, payload.mwm);
        process.send(pow);
    }

    if (payload.job === 'batchedPow') {
        const pow = await EntangledNode.powBundleFunc(
            payload.trytes,
            payload.trunkTransaction,
            payload.branchTransaction,
            payload.mwm,
        );
        process.send(pow);
    }

    if (payload.job === 'gen') {
        const address = await EntangledNode.genAddressTritsFunc(payload.seed, payload.index, payload.security);
        process.send(address);
    }
});

const Entangled = {
    powFn: async (trytes, mwm) => {
        return await exec(JSON.stringify({ job: 'pow', trytes, mwm }));
    },
    batchedPowFn: async (trytes, trunkTransaction, branchTransaction, mwm) => {
        return await exec(JSON.stringify({ job: 'batchedPow', trytes, trunkTransaction, branchTransaction, mwm }));
    },
    genFn: async (seed, index, security) => {
        return await exec(JSON.stringify({ job: 'gen', seed, index, security }));
    },
};

module.exports = Entangled;
