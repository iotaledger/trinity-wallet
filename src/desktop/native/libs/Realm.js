import Realm from 'realm';

// Don't report Realm JS analytics data
// @see https://github.com/realm/realm-js/blob/master/lib/submit-analytics.js#L28
process.env.REALM_DISABLE_ANALYTICS = 'true';

export default Realm;
