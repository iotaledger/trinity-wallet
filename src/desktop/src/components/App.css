@import-normalize;

@custom-media --xs (min-width: 0px); /* default */
@custom-media --sm (max-width: 480px);
@custom-media --md (max-width: 768px);
@custom-media --lg (max-width: 1024px);
@custom-media --xl (max-width: 1440px);

:root {
    --color-text: white;
    --color-yellow: #f7d002;
    --color-blue: #8bd4ff;
    --color-red: #ff6c69;
    --color-green: #9dffaf;
    --color-darkgreen: green;

    --color-label: var(--color-yellow);
    --color-button-warning: var(--color-yellow);
    --color-button-danger: var(--color-red);
    --color-button-success: var(--color-green);
    --color-button-extra: var(--color-blue);
    --color-button-cta: var(--color-darkgreen);
    --color-button-default: white;

    /*--font-size-default: 1.8vh;*/
    --font-size-default: 14px;
    --font-size-small: 12px;
    --font-size-title: 20px;

    --z-index-modal: 2048;
    --z-index-overlay: 1024;
    --z-index-dropdown: 512;
    --z-index-elevated: 1;
}

* {
    &,
    &:before,
    &:after {
        box-sizing: border-box;
    }
}

html,
body {
    margin: 0;
    padding: 0;
    height: 100%;
}

body {
    background-image: url(../../../shared/images/bg-green.jpg);
    background-repeat: no-repeat;
    background-size: 100% 100%;
    background-attachment: fixed;
    color: var(--color-text);
    font-family: arial;
    font-size: var(--font-size-default);
    line-height: 1.5;
}

h1 {
    font-size: var(--font-size-title);
}

label {
    color: var(--color-label);
    display: block;
    margin-bottom: 8px;
}

select {
    &:focus {
        outline: 0;
    }
    background: rgba(0, 0, 0, 0.2);
    border: 0;
    padding: 16px 4px;
    width: 100%;
    border-bottom: 2px solid white;
    color: white;
}

option {
    background: transparent;
    color: black;
    padding: 4px;
}

input {
    box-sizing: border-box;

    &:focus {
        outline: none;
    }

    &[type='text'],
    &[type='email'],
    &[type='tel'],
    &[type='password'] {
        background: rgba(0, 0, 0, 0.2);
        border: 0;
        border-bottom: 2px solid white;
        color: white;
        padding: 16px;
        width: 100%;
    }
}

textarea {
    background: rgba(0, 0, 0, 0.2);
    border: 0;
    border-bottom: 2px solid white;
    box-sizing: border-box;
    color: white;
    padding: 16px;
    width: 100%;
    &:focus {
        outline: none;
    }
}
