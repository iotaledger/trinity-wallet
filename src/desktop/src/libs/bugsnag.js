/* global Electron */
import React from 'react';
import bugsnag from '@bugsnag/js';
import bugsnagReact from '@bugsnag/plugin-react';

import settings from '../../package.json';

const bugsnagKey = process.env.BUGSNAG_KEY;

export const bugsnagClient = bugsnagKey.length
    ? bugsnag({
          apiKey: bugsnagKey,
          appVersion: settings.version,
          interactionBreadcrumbsEnabled: false,
          collectUserIp: false,
          user: { id: typeof Electron === 'object' ? Electron.getUuid() : null },
      })
    : {
          notify: (err) => {
              console.error(err); /* eslint-disable-line no-console */
          },
      };

if (bugsnagKey.length) {
    bugsnagClient.use(bugsnagReact, React);
}

export const ErrorBoundary = bugsnagKey.length ? bugsnagClient.getPlugin('react') : React.Fragment;
