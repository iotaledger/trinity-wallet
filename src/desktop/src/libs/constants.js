import { version } from '../../package.json';

/**
 * Use different account aliases for Relase Candidate versions
 */
export const __RC__ = version.toLowerCase().indexOf('rc') > 0 && process.env.NODE_ENV !== 'test' ? '-rc' : '';

/**
 * Main account encryption key alias
 */
export const ALIAS_MAIN = `Trinity${__RC__}`;

/**
 * Seed account encryption key alias
 */
export const ALIAS_ACCOUNT = `account${__RC__}`;

/**
 * Realm database encryption key alias
 */
export const ALIAS_REALM = `realm_enc_key${__RC__}`;

/**
 * Legacy credentials key alias
 */
export const ALIAS_LEGACY = `moonpay_credentials${__RC__}`;
