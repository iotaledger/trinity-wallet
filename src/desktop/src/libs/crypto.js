/* global Electron */
import { MAX_SEED_LENGTH } from 'libs/iota/utils';

// Prefix for seed account titles stored in the vault
const ACC_PREFIX = 'account';

// Title of main vault entry  containing 2fa keys
export const ACC_MAIN = 'Trinity';
// Maximum allowed account title
export const MAX_ACC_LENGTH = 250;

/**
 * Create random byte array
 * @param {Number} Length - Random number array length.
 * @param {Number} Max -  Random byte max range
 * @returns {Array} Random number array
 */
function randomBytes(size, max) {
    if (size !== parseInt(size, 10) || size < 0) {
        return false;
    }

    const rawBytes = new Uint8Array(size);

    const bytes = global.crypto.getRandomValues(rawBytes);

    for (let i = 0; i < bytes.length; i++) {
        while (bytes[i] >= 256 - (256 % max)) {
            bytes[i] = randomBytes(1, max)[0];
        }
    }

    return Array.from(bytes);
}

/**
 * Create random seed
 * @param {Number} length - Random seed length
 * @returns {Array} Random byte array seed
 */
export const createRandomSeed = (length = MAX_SEED_LENGTH) => {
    return randomBytes(length, 27);
};

/**
 * Encrypt plain text
 * @param {Any} Content - Content to encrypt
 * @param {String} Password - Plain text password for encryption
 * @returns {String} Ecnrypted initialization vector + content
 */
const encrypt = async (contentPlain, passwordPlain) => {
    const content = new TextEncoder().encode(JSON.stringify(contentPlain));

    const password = new TextEncoder().encode(passwordPlain);
    const passwordhash = await crypto.subtle.digest('SHA-256', password);

    const iv = crypto.getRandomValues(new Uint8Array(12));

    const algorithm = { name: 'AES-GCM', iv: iv };

    const key = await crypto.subtle.importKey('raw', passwordhash, algorithm, false, ['encrypt']);

    const cipherBuffer = await crypto.subtle.encrypt(algorithm, key, content);
    const cipherArray = new Uint8Array(cipherBuffer);
    const cipherHex = cipherArray.toString();

    const ivHex = iv.toString();

    return `${ivHex}|${cipherHex}`;
};

/**
 * Decrypt cyphertext
 * @param {String} Content - Encrypted initialization vector + content
 * @param {String} Password - Plain text password for decryption
 * @returns {Any} Derypted content
 */
const decrypt = async (cipherText, passwordPlain) => {
    const cipherParts = cipherText.split('|');

    if (cipherParts.length !== 2 || typeof passwordPlain !== 'string' || !passwordPlain.length) {
        throw new Error('Wrong password');
    }
    try {
        const password = new TextEncoder().encode(passwordPlain);
        const passwordHash = await crypto.subtle.digest('SHA-256', password);

        const ivArray = cipherParts[0].split(',');
        const iv = Uint8Array.from(ivArray);

        const algorithm = { name: 'AES-GCM', iv: iv };

        const key = await crypto.subtle.importKey('raw', passwordHash, algorithm, false, ['decrypt']);

        const cipherArray = cipherParts[1].split(',');
        const cipher = Uint8Array.from(cipherArray);

        const plainBuffer = await crypto.subtle.decrypt(algorithm, key, cipher);
        const plainText = new TextDecoder().decode(plainBuffer);

        return JSON.parse(plainText);
    } catch (err) {
        throw new Error('Wrong password');
    }
};

/**
 * Check for valid vault password
 * @param {String} Password - Plain text password for decryption
 * @returns {String} Two factor authentication key
 */
export const vaultAuth = async (password) => {
    const vault = await Electron.readKeychain(ACC_MAIN);
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const decryptedVault = await decrypt(vault, password);
        if (decryptedVault.twoFaKey) {
            return decryptedVault.twoFaKey;
        }
        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Clear the vault
 * @param {String} Password - Plain text password for decryption
 * @param {Boolean} Reset - Should the vault be reset without authentication
 * @returns {Boolean} True if vault cleared
 */
export const clearVault = async (password, reset) => {
    if (!reset) {
        await vaultAuth(password);
    }

    const vault = await Electron.listKeychain();

    const accounts = Object.keys(vault);

    for (let i = 0; i < accounts.length; i++) {
        await Electron.removeKeychain(vault[i].account);
    }

    return true;
};

/**
 * Update vault password
 * @param {String} PasswordCurrent - Current plain text password
 * @param {String} PasswordNew - New plain text password
 * @returns {Boolean} True if password updated
 */
export const updatePassword = async (passwordCurrent, passwordNew) => {
    const vault = await Electron.listKeychain();
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const accounts = Object.keys(vault);

        if (!accounts.length) {
            return true;
        }

        for (let i = 0; i < accounts.length; i++) {
            const account = vault[i];

            const decryptedVault = await decrypt(account.password, passwordCurrent);
            const encryptedVault = await encrypt(decryptedVault, passwordNew);

            Electron.setKeychain(account.account, encryptedVault);
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Get seed from keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} SeedName - Seed name to retreive from keychain
 * @param {Boolean} PlainText - Should the seed be returned in plain ACII text
 * @returns {Array} Derypted seed
 */
export const getSeed = async (password, seedName, plainText) => {
    const seedNameHash = await hashSeedName(seedName);

    const vault = await Electron.readKeychain(seedNameHash);
    if (!vault) {
        throw new Error('Incorrect seed name');
    }
    try {
        const decryptedVault = await decrypt(vault, password);
        return plainText ? seedToHex(decryptedVault) : decryptedVault;
    } catch (err) {
        throw err;
    }
};

/**
 * Save seed to keychain
 * @param {String} Password - Plain text password for encryption
 * @param {String} SeedName - Seed name to set to keychain
 * @param {Array} Seed - Seed array
 * @returns {Boolean} True if seed is saved to keychain
 */
export const setSeed = async (password, seedName, seed) => {
    try {
        const seedNameHash = await hashSeedName(seedName);
        const vault = await encrypt(Array.from(seed), password);

        Electron.setKeychain(seedNameHash, vault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Set Two-Factor authentication key
 * @param {String} Password - Plain text password for decryption
 * @param {String} Key - Two-factor authentication key
 * @returns {Boolean} True if Two-Factor key is set
 */
export const setTwoFA = async (password, key) => {
    try {
        const vault = await Electron.readKeychain(ACC_MAIN);
        const decryptedVault = vault === null ? {} : await decrypt(vault, password);

        if (key) {
            decryptedVault.twoFaKey = key;
        } else {
            delete decryptedVault.twoFaKey;
        }

        const updatedVault = await encrypt(decryptedVault, password);

        Electron.setKeychain('Trinity', updatedVault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Remove seed from keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} SeedName - Seed name to remove from keychain
 * @returns {Boolean} True seed is removed
 */
export const removeSeed = async (password, seedName) => {
    try {
        const seedNameHash = await hashSeedName(seedName);

        const isRemoved = await Electron.removeKeychain(seedNameHash);
        if (!isRemoved) {
            throw new Error('Incorrect seed name');
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Rename seed in keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} OldName - Current seed name
 * @param {String} NewName - New seed name
 * @returns {Boolean} True if seed is renamed
 */
export const renameSeed = async (password, seedName, newSeedName) => {
    const seedNameHash = await hashSeedName(seedName);
    const newNameHash = await hashSeedName(newSeedName);

    const vault = await Electron.readKeychain(seedNameHash);

    if (!vault) {
        throw new Error('Incorrect seed name');
    }

    try {
        await decrypt(vault, password);
        Electron.removeKeychain(seedNameHash);
        Electron.setKeychain(newNameHash, vault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Unique seed check
 * @param {String} Password - Plain text password for decryption
 * @param {Array} Seed - Seed to check
 * @returns {Boolean} True if seed is unique
 */
export const uniqueSeed = async (password, seed) => {
    const vault = await Electron.listKeychain();
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const accounts = vault.filter((acc) => acc.account !== ACC_MAIN);

        for (let i = 0; i < accounts.length; i++) {
            const account = vault[i];
            const vaultSeed = await decrypt(account.password, password);
            if (vaultSeed.length === seed.length && seed.every((v, x) => v % 27 === vaultSeed[x] % 27)) {
                return false;
            }
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Hash text using SHA-256
 * @param {String} Password - plain text to hash
 * @returns {String} SHA-256 hash
 */
export const sha256 = async (inputPlain) => {
    if (typeof inputPlain !== 'string' || inputPlain.length < 1) {
        return false;
    }
    const input = new TextEncoder().encode(inputPlain);
    const hash = await crypto.subtle.digest('SHA-256', input);
    const plainHash = bufferToHex(hash);

    return plainHash;
};

/**
 * Hash seed name using SHA-256
 * @param {String} SeedName - plain text to hash
 * @returns {String} SHA-256 hash
 */
const hashSeedName = async (seedName) => {
    const prefixName = `${ACC_PREFIX}-${seedName}`;
    const hash = await sha256(prefixName);
    return hash;
};

/**
 * Check for a valid activation code
 * @param {String} code - Target activation code
 * @param {String} uuid - UUID of the machine
 * @returns {Boolean} True if activation code correct
 */
export const checkActivationCode = async (code, uuid) => {
    const key = 'LURGzCPEHqhjvYLwAJXRv5Fc';
    const hash = await sha256(key + uuid);
    return code === hash;
};

/**
 * Convert byte seed array to string
 * @param {Array} seed - Target seed array
 * @returns {String} Plain text seed string
 */
const seedToHex = (bytes) => {
    return Array.from(bytes)
        .map((byte) => byteToChar(byte % 27))
        .join('');
};

/**
 * Convert single character byte to string
 * @param {Number} byte - Input byte
 * @returns {String} Output character
 */
export const byteToChar = (byte) => {
    return '9ABCDEFGHIJKLMNOPQRSTUVWXYZ'.charAt(byte % 27);
};

/**
 * Convert buffer to plain text
 * @param {Buffer} - Input buffer
 * @returns {String} Output string
 */
const bufferToHex = (buffer) => {
    const hexCodes = [];
    const view = new DataView(buffer);
    for (let i = 0; i < view.byteLength; i += 4) {
        const value = view.getUint32(i);
        const stringValue = value.toString(16);
        const padding = '00000000';
        const paddedValue = (padding + stringValue).slice(-padding.length);
        hexCodes.push(paddedValue);
    }

    return hexCodes.join('');
};
