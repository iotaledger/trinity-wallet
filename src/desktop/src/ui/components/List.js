import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { formatValue, formatUnit, round } from 'libs/util';
import { convertFromTrytes } from 'libs/iota/utils';
import { formatTime, formatModalTime, convertUnixTimeToJSDate } from 'libs/dateUtils';
import Modal from 'ui/components/modal/Modal';
import Button from 'ui/components/Button';
import Clipboard from 'ui/components/Clipboard';
import Icon from 'ui/components/Icon';

import withListData from 'containers/components/List';

import css from './list.css';

/**
 * Transaction history list component
 */
class List extends React.PureComponent {
    static propTypes = {
        /** Transaction history */
        transfers: PropTypes.array.isRequired,
        /** Set active history item
         * @param {Number} index - Current item index
         */
        setItem: PropTypes.func.isRequired,
        /** Current active history item */
        currentItem: PropTypes.number,
        /** Receive address list */
        addresses: PropTypes.array.isRequired,
        /** Translation helper
         * @param {string} translationString - locale string identifier to be translated
         * @ignore
         */
        t: PropTypes.func.isRequired,
    };

    state = {
        filter: 'All',
    };

    switchFilter(filter) {
        this.setState({
            filter: filter,
        });
    }

    render() {
        const { transfers, addresses, setItem, currentItem, t } = this.props;
        const { filter } = this.state;

        const filters = ['All', 'Sent', 'Received', 'Pending'];

        const activeTransfers = currentItem ? transfers[parseInt(currentItem)] : null;
        const activeTransfer = activeTransfers ? activeTransfers[0] : null;

        return (
            <React.Fragment>
                <nav className={css.nav}>
                    {filters.map((item) => {
                        return (
                            <a
                                key={item}
                                onClick={() => this.switchFilter(item)}
                                className={classNames(
                                    filter === item ? css.active : null,
                                    !transfers || !transfers.length ? css.disabled : null,
                                )}
                            >
                                {item}
                            </a>
                        );
                    })}
                </nav>
                <hr />
                <div className={css.list}>
                    {transfers && transfers.length ? (
                        transfers.map((transferRow, key) => {
                            const transfer = transferRow[0];
                            const isReceived = addresses.includes(transfer.address);
                            const isConfirmed = transfer.persistence;

                            if (
                                (filter === 'Sent' && isReceived) ||
                                (filter === 'Received' && !isReceived) ||
                                (filter === 'Pending' && isConfirmed)
                            ) {
                                return null;
                            }

                            return (
                                <a
                                    key={key}
                                    onClick={() => setItem(key)}
                                    className={classNames(
                                        isReceived ? css.received : css.sent,
                                        isConfirmed ? css.confirmed : css.pending,
                                    )}
                                >
                                    <div>
                                        <Icon icon={isReceived ? 'receive' : 'send'} size={16} />
                                        <span>{formatTime(convertUnixTimeToJSDate(transfer.timestamp))}</span>
                                        <strong>
                                            {!isConfirmed ? t('pending') : isReceived ? t('received') : t('sent')}
                                        </strong>
                                        <span>{`${round(formatValue(transfer.value))} ${formatUnit(
                                            transfer.value,
                                        )}`}</span>
                                    </div>
                                </a>
                            );
                        })
                    ) : (
                        <p className={css.empty}>No recent history</p>
                    )}
                </div>
                {activeTransfer !== null ? (
                    <Modal isOpen onClose={() => setItem(null)}>
                        <div className={css.historyItem}>
                            <header
                                className={classNames(
                                    addresses.includes(activeTransfer.address) ? css.received : css.sent,
                                    activeTransfer.persistence ? css.confirmed : css.pending,
                                )}
                            >
                                <p>
                                    <strong>
                                        {!activeTransfer.persistence
                                            ? t('pending')
                                            : addresses.includes(activeTransfer.address) ? t('received') : t('sent')}
                                    </strong>{' '}
                                    <span>{`${round(formatValue(activeTransfer.value))} ${formatUnit(
                                        activeTransfer.value,
                                    )}`}</span>
                                </p>
                                <small>{formatModalTime(convertUnixTimeToJSDate(activeTransfer.timestamp))}</small>
                            </header>
                            <h6>Bundle Hash:</h6>
                            <p>
                                <Clipboard
                                    text={activeTransfer.bundle}
                                    title={t('history:bundleHashCopied')}
                                    success={t('history:bundleHashCopiedExplanation')}
                                />
                            </p>
                            <h6>Addresses</h6>
                            <div className={css.scroll}>
                                <ul>
                                    {activeTransfers.map((tx, key) => {
                                        return (
                                            <li key={key}>
                                                <p>
                                                    <Clipboard
                                                        text={tx.address}
                                                        title={t('history:addressCopied')}
                                                        success={t('history:addressCopiedExplanation')}
                                                    />
                                                </p>
                                                <strong>{`${round(formatValue(tx.value))} ${formatUnit(
                                                    tx.value,
                                                )}`}</strong>
                                            </li>
                                        );
                                    })}
                                </ul>
                            </div>
                            <h6>Message</h6>
                            <p>{convertFromTrytes(activeTransfer.signatureMessageFragment)}</p>
                            <footer>
                                <Button onClick={() => setItem(null)} variant="primary">
                                    {t('back')}
                                </Button>
                            </footer>
                        </div>
                    </Modal>
                ) : null}
            </React.Fragment>
        );
    }
}

export default withListData(List);
