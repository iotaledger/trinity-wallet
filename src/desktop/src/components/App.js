import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { persistStore } from 'redux-persist';
import { withRouter } from 'react-router-dom';
import store from 'store';
import i18next from 'libs/i18next';
import Loading from 'components/UI/Loading';
import Onboarding from 'components/Layout/Onboarding';
import Main from 'components/Layout/Main';
import Notifications from 'components/UI/Notifications';

import './App.css';

// should be used later:

// class ErrorBoundary extends React.Component {
//     static propTypes = {
//         children: PropTypes.node,
//     };
//
//     state = {};
//
//     componentDidCatch(error) {
//         this.setState(() => ({
//             error,
//         }));
//     }
//     render() {
//         if (this.state.error) {
//             return <p>{this.state.error.message}</p>;
//         }
//         return this.props.children;
//     }
// }

class App extends React.Component {
    static propTypes = {
        settings: PropTypes.shape({
            locale: PropTypes.string.isRequired,
            fullNode: PropTypes.string.isRequired,
        }).isRequired,
        app: PropTypes.shape({
            isOnboardingCompleted: PropTypes.bool.isRequired,
        }).isRequired,
    };

    state = {
        initialized: false,
    };

    componentWillMount() {
        persistStore(store, { blacklist: ['tempAccount', 'notifications', 'seeds'] }, () => {
            // setTimeout(
            //     () =>
            //         this.setState(() => ({
            //             initialized: true,
            //         })),
            //     2500,
            // );
            // TODO: re-add timeout to avoid flashes of the loading spinner. temporarily disabled for easier debugging
            this.setState(() => ({
                initialized: true,
            }));
        });
    }

    // TODO: this is not working on windows. Investigate why
    // componentDidMount() {
    //     const { remote } = require('electron');
    //     const keytar = remote.require('keytar');
    //     // const keytar = require('keytar');
    //     keytar
    //         .setPassword('iotaWallet', 'Main Wallet', 'AAABBBCCCC999')
    //         .then(() => {
    //             console.log('SAVED SEED');
    //         })
    //         .catch(err => {
    //             console.log('ERROR WHILE SAVING', err);
    //         });
    // }

    componentWillReceiveProps(nextProps) {
        if (nextProps.settings.locale !== this.props.settings.locale) {
            i18next.changeLanguage(nextProps.settings.locale);
        }
    }

    componentDidCatch(error) {
        this.setState(() => ({
            error,
        }));
    }

    render() {
        const { app } = this.props;

        if (this.state.initialized === false) {
            return <Loading />;
        }

        return (
            <div>
                {this.state.error && <p>{this.state.error.message}</p>}
                <Notifications />
                {app.isOnboardingCompleted ? <Main /> : <Onboarding />}
            </div>
        );
    }
}

const mapStateToProps = state => ({
    settings: state.settings,
    app: state.app,
});

export default withRouter(connect(mapStateToProps)(App));
