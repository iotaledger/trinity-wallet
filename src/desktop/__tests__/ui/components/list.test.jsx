import React from 'react';
import { shallow } from 'enzyme';

import { ListComponent as List } from 'ui/components/List';

import * as dateLib from 'libs/date';
dateLib.formatTime = () => 'DD/MM/YYYY';

const props = {
    isBusy: false,
    mode: 'Standard',
    isLoading: false,
    currentlyPromotingBundleHash: '',
    isRetryingFailedTransaction: false,
    hideEmptyTransactions: false,
    updateAccount: jest.fn(),
    toggleEmptyTransactions: jest.fn(),
    /** Transaction history */
    transactions: [
        {
            attachmentTimestamp: 1546934663241,
            bundle: 'ABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHI',
            incoming: false,
            inputs: [],
            message: 'Lorem ipsum',
            persistence: true,
            timestamp: 1546934661,
            transferValue: 0,
        },
        {
            attachmentTimestamp: 1546934663242,
            bundle: 'BCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJ',
            incoming: true,
            inputs: [],
            message: 'Dolor sit amet',
            persistence: false,
            timestamp: 1546934661,
            transferValue: 500000,
        },
        {
            attachmentTimestamp: 1546934663242,
            bundle: 'CDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJK',
            incoming: true,
            inputs: [],
            message: 'Lorem ipsum',
            persistence: true,
            timestamp: 1546934661,
            transferValue: 500000,
        },
        {
            attachmentTimestamp: 1546934663243,
            bundle: 'DEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKI',
            incoming: false,
            inputs: [],
            message: 'Lorem ipsum',
            persistence: true,
            timestamp: 1546934661,
            transferValue: 0,
        },
        {
            attachmentTimestamp: 1546934663243,
            bundle: 'EFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKIL',
            incoming: false,
            inputs: [],
            message: 'Lorem ipsum',
            persistence: false,
            timestamp: 1546934661,
            transferValue: 100000,
        },
        {
            attachmentTimestamp: 1546934663243,
            bundle: 'FGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKILM',
            incoming: false,
            inputs: [],
            message: 'Lorem ipsum',
            persistence: false,
            timestamp: 1546934661,
            transferValue: 100000,
        },
    ],
    failedHashes: {},
    promoteTransaction: jest.fn(),
    retryFailedTransaction: jest.fn(),
    setItem: jest.fn(),
    t: (str) => str,
    accountMeta: {
        type: 'keychain',
    },
    password: {},
};

describe('List component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<List {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('List all history items', () => {
        const wrapper = shallow(<List {...props} />);

        expect(wrapper.find('.list a')).toHaveLength(6);
    });

    test('Filter received items', () => {
        const wrapper = shallow(<List {...props} />);

        wrapper.setState({ filter: 'Received' });
        expect(wrapper.find('.list a')).toHaveLength(1);
    });

    test('Filter sent items', () => {
        const wrapper = shallow(<List {...props} />);

        wrapper.setState({ filter: 'Sent' });
        expect(wrapper.find('.list a')).toHaveLength(2);
    });

    test('Filter pending items', () => {
        const wrapper = shallow(<List {...props} />);

        wrapper.setState({ filter: 'Pending' });
        expect(wrapper.find('.list a')).toHaveLength(3);
    });

    test('Filter items by value search', () => {
        const wrapper = shallow(<List {...props} />);

        wrapper.setState({ search: '500000' });
        expect(wrapper.find('.list a')).toHaveLength(2);
    });

    test('Filter items by message search', () => {
        const wrapper = shallow(<List {...props} />);

        wrapper.setState({ search: 'dolor' });
        expect(wrapper.find('.list a')).toHaveLength(1);
    });

    test('Display single transaction', () => {
        const mockProps = Object.assign({}, props, {
            currentItem: 'ABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHIJKLMONPRSTUVXYZABCDEFGHI',
        });
        const wrapper = shallow(<List {...mockProps} />);

        expect(wrapper.find('.popup').hasClass('on')).toBeTruthy();
    });
});
