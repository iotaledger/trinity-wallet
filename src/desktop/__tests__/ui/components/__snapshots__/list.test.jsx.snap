// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Render the component 1`] = `
<Fragment>
  <nav
    className="nav"
  >
    <ul>
      <a
        key="active"
        onClick={[Function]}
      >
        all
         
        <small>
          (
          6
          )
        </small>
        <Icon
          icon="chevronDown"
          size={8}
        />
      </a>
      <li>
        <a
          className="active"
          key="All"
          onClick={[Function]}
        >
          global:all
           (
          6
          )
        </a>
        <a
          className=""
          key="Sent"
          onClick={[Function]}
        >
          sent
           (
          2
          )
        </a>
        <a
          className=""
          key="Received"
          onClick={[Function]}
        >
          received
           (
          1
          )
        </a>
        <a
          className=""
          key="Pending"
          onClick={[Function]}
        >
          pending
           (
          3
          )
        </a>
        <div>
          <a
            className="checkbox off"
            onClick={[Function]}
          >
            history:hideZeroBalance
          </a>
        </div>
      </li>
    </ul>
    <div
      className="search"
    >
      <input
        className={null}
        onChange={[Function]}
        placeholder="history:typeHelp"
        value=""
      />
      <div
        onClick={[Function]}
      >
        <Icon
          icon="search"
          size={20}
        />
      </div>
    </div>
    <a
      className="refresh"
      onClick={[Function]}
    >
      <Icon
        icon="sync"
        size={24}
      />
    </a>
  </nav>
  <hr />
  <div
    className="list"
  >
    <Scrollbar>
      <a
        className="confirmed sent"
        key="0"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="minus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            sent
          </span>
          <span>
            0 i
          </span>
        </div>
      </a>
      <a
        className="pending received"
        key="1"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="plus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            receiving
          </span>
          <span>
            +
            500 Ki
          </span>
        </div>
      </a>
      <a
        className="confirmed received"
        key="2"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="plus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            received
          </span>
          <span>
            +
            500 Ki
          </span>
        </div>
      </a>
      <a
        className="confirmed sent"
        key="3"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="minus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            sent
          </span>
          <span>
            0 i
          </span>
        </div>
      </a>
      <a
        className="pending sent"
        key="4"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="minus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            sending
          </span>
          <span>
            -
            100 Ki
          </span>
        </div>
      </a>
      <a
        className="pending sent"
        key="5"
        onClick={[Function]}
      >
        <div>
          <Icon
            icon="minus"
            size={14}
          />
          <span>
            DD/MM/YYYY
          </span>
          <span>
            sending
          </span>
          <span>
            -
            100 Ki
          </span>
        </div>
      </a>
    </Scrollbar>
  </div>
  <div
    className="popup"
    onClick={[Function]}
  >
    <div />
  </div>
</Fragment>
`;
